### ISSUE TITLE
C# Client hangs when invoking a hub method and waiting on the invocation

### ISSUE THREAD

**Issue by NTaylorMullen** [SignalR/SignalR#2153]
``` C#
var connection = new HubConnection(...);
var proxy = connection.CreateProxy("EchoHub");

proxy.On("Echo", value => 
{
    proxy.Invoke("Echo", value + 1).Wait();
});

connection.Start().Wait();
proxy.Invoke("Echo", 1).Wait();
```

Causes the client to hang due to a lock already been taken on the client

**Comment by jjxtra** [comment]
I get hangs regularly on Invoke with .NET 4.7.1 / .net core 2.1, is this still a known issue?

**Comment by chartierpw** [comment]
I'm having a hang issue with Microsoft.AspNet.SignalR.Client v2.4, wondering if they're related?  I call a hub method from android client and it hangs.  Client code is:

        public List<string> ConnectedUsers()
        {
            // Hangs on this line
            List<string> Users = _proxy.Invoke<List<string>>("getConnectedUsers").Result;

            return Users;
        }

Hub method code is:

        public List<string> getConnectedUsers()
        {
            List<string> UserNames = new List<string>();

            foreach (var ConnectedUser in Users)
            {
                UserNames.Add(ConnectedUser.Name );
            }
            
           return UserNames;
        }

The call works with javascript, IE;
Clicking the get users button executes the following:

                    $('#getusers').click
                    (
                        function () 
                        {
                            chat.server.getConnectedUsers().done(function (result) {
                                alert(result);
                            });
                        }
                    );

This works fine.

**Comment by davidfowl** [comment]
You should avoid using .Result read this https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#avoid-using-taskresult-and-taskwait

**Comment by chartierpw** [comment]
@davidfowl 
Thank you!  I've been working on this for a week and never once saw this document.  I changed my code in accordance with the information in the document and now it works as expected.  For those researching the same issue, I changed my code to:

On the client:
From:

    public List<string> ConnectedUsers()
    {
        // Hangs on this line
        List<string> Users = _proxy.Invoke<List<string>>("getConnectedUsers").Result;

        return Users;
    }

To:

        public async Task <List<string>> ConnectedUsers()
        {

            List<string> Users = await _proxy.Invoke<List<string>>("getConnectedUsers");

            return Users;
        }

The call to the ConnectedUsers function in Client.cs was changed as well:
From:

                List<string> userList = client.ConnectedUsers();


To:
                List<string> userList = await client.ConnectedUsers();

No changes to the hub code were necessary.

**Comment by chartierpw** [comment]
@davidfowl
I wonder if you could assist me with another issue I'm having.  I'd like to be able to access the connection to the hub from anywhere in the application.  After authenticating the user I connect to the hub, wire up an event to receive messages and continue on to the main view of the application.  Messages sent to the hub by other users are received and displayed with no problem, but now I need to be able to respond to/and send messages over the same connection from other areas of the application without having to reconnect.  I'm using the latest version of Visual Studio 2017 (15.9.11) with the latest version of Xamarin (2.12.3.82).  Any assistance would be greatly appreciated.  Thanks.

**Comment by davidfowl** [comment]
Asking questions like this should start on stack overflow

**Comment by chartierpw** [comment]
I was hoping for a direct communication.  An earlier search of Stack 
Overflow did not return anything useful and experienced users of SignalR 
seem to be at a premium.  I've asked several questions regarding 
Microsoft.AspNetCore.SignalR which I was never able to get to work and 
Microsoft.AspNet.SignalR which I was able to get to work.

For example 
https://stackoverflow.com/questions/55403257/connection-to-signalr-hub-fails-with-negotiate-404-not-found

I have no faith in Stack Overflow where it concerns SignalR, but I will try 
again.  Thanks.

R/
Prescott ...
-----Original Message-----

**Comment by davidfowl** [comment]
We also look at stackoverflow. GitHub isn’t great for open ended questions. Start there and I’ll look at the question and if you don’t get feedback, you can file a new issue. Commenting on closed issues is a sure way to get ignored

**Comment by chartierpw** [comment]
Of course, just very frustrated with getting nowwhere ... lol ... been on 
this for 2 weeks.  I just finished posting the question.

https://stackoverflow.com/questions/55779814/send-message-to-signalr-hub-from-any-xamarin-visual-studio-2017-android-activi

Thanks for the help and advice.

R/
Prescott ....
-----Original Message-----

**Comment by NanaseRuri** [comment]
> You should avoid using .Result read this https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#avoid-using-taskresult-and-taskwait

Thanks. It works perfectly by using await instead of Wait().
